client/components/command-palette/test/index.tsx

```diff
@@ -0,0 +1,183 @@
+			expect( screen.getByText( 'Get help' ) ).toBeInTheDocument();
+		} );
+	} );
+
+	it( 'should close the command palette when Escape key is pressed', () => {
+		renderCommandPalette();
+
+		waitFor( () => {
+			expect( screen.getByPlaceholderText( 'Search for commands' ) ).toBeInTheDocument();
+			fireEvent.keyDown( document, { key: 'Escape' } );
+		} );
+
+		waitFor( () => {
+			expect( screen.queryByPlaceholderText( 'Search for commands' ) ).toBeNull();
+		} );
+	} );
+
+	it( 'should close the command palette when clicking outside the modal overlay', () => {
+		renderCommandPalette();
+
+		waitFor( () => {
+			expect( screen.getByPlaceholderText( 'Search for commands' ) ).toBeInTheDocument();
+			const overlayElement = document.querySelector( '.commands-command-menu__overlay' );
+			overlayElement && fireEvent.click( overlayElement );
+		} );
+
+		waitFor( () => {
+			expect( screen.queryByPlaceholderText( 'Search for commands' ) ).toBeNull();
+		} );
+	} );
+
+	it( 'should close the palette when you select a specific command', () => {
+		renderCommandPalette();
+
+		waitFor( () => {
+			expect( screen.getByPlaceholderText( 'Search for commands' ) ).toBeInTheDocument();
+			const getHelpCommand = screen.getByText( 'Get help' );
+			fireEvent.click( getHelpCommand );
+		} );
+
+		waitFor( () => {
+			expect( screen.queryByPlaceholderText( 'Search for commands' ) ).toBeNull();
+		} );
+	} );
+
+	it( 'should return only "Get help" command as it matches label and "Send feedback" as it matches searchLabel; other commands are hidden', () => {
+		renderCommandPalette();
+
+		waitFor( () => {
+			expect( screen.getByPlaceholderText( 'Search for commands' ) ).toBeInTheDocument();
+			fireEvent.change( screen.getByPlaceholderText( 'Search for commands' ), {
+				target: { value: 'help' },
+			} );
+		} );
+
+		waitFor( () => {
+			expect( screen.getByText( 'Get help' ) ).toBeInTheDocument();
+			expect( screen.getByText( 'Send feedback' ) ).toBeInTheDocument();
+			expect( screen.queryByText( 'Clear cache' ) ).toBeNull();
+			expect( screen.queryByText( 'Enable edge cache' ) ).toBeNull();
+		} );
+	} );
+
+	it( 'should return "No results found" when there is no match for search', () => {
+		renderCommandPalette();
+
+		waitFor( () => {
+			expect( screen.getByPlaceholderText( 'Search for commands' ) ).toBeInTheDocument();
+			fireEvent.change( screen.getByPlaceholderText( 'Search for commands' ), {
+				target: { value: 'blue' },
+			} );
+		} );
+
+		waitFor( () => {
+			expect( screen.getByText( 'No results found.' ) ).toBeInTheDocument();
+			expect( screen.getByText( 'Send feedback' ) ).toBeNull();
+			expect( screen.queryByText( 'Clear cache' ) ).toBeNull();
+			expect( screen.queryByText( 'Enable edge cache' ) ).toBeNull();
+		} );
+	} );
+
+	it( 'should navigate to a nested command and back to root commands', () => {
+		renderCommandPalette();
+
+		waitFor( () => {
```

Comment:
Just a sidenote: the usage of the `waitFor` might look strange here without `async` but I had to use it to ensure that the DOM updates caused by the simulated events e.g. clicking the button have been processed before making new assertions. This helps to prevent timing issues. 

I tried removing it but it causes the test to time out so the best option I found is to use `waitFor` in this way to ensure the sequence of events runs as expected. Happy to modify it though if you have better suggestions.

Reply:
@katinthehatsite , I think the `waitFor` is not working as expected. I tried changing the values on `getByPlaceholderText` and `getByText` and the tests pass without any error. I don't have a solution yet.

I tried returning the render values:
```
const renderCommandPalette = () => {
 ...
		return render(
			<Provider store={ store }>
				<CommandPalette />
			</Provider>
		);
```

And then grabbing the properties in each test:
```
const { getByText, container, rerender, debug } = renderCommandPalette();
```

But the markup I receive is this:
```
 <body>
        <p
          class="a11y-speak-intro-text"
          hidden="hidden"
          id="a11y-speak-intro-text"
          style="position: absolute;margin: -1px;padding: 0;height: 1px;width: 1px;overflow: hidden;clip: rect(1px, 1px, 1px, 1px);-webkit-clip-path: inset(50%);clip-path: inset(50%);border: 0;word-wrap: normal !important;"
        >
          Notifications
        </p>
        <div
          aria-atomic="true"
          aria-live="assertive"
          aria-relevant="additions text"
          class="a11y-speak-region"
          id="a11y-speak-assertive"
          style="position: absolute;margin: -1px;padding: 0;height: 1px;width: 1px;overflow: hidden;clip: rect(1px, 1px, 1px, 1px);-webkit-clip-path: inset(50%);clip-path: inset(50%);border: 0;word-wrap: normal !important;"
        />
        <div
          aria-atomic="true"
          aria-live="polite"
          aria-relevant="additions text"
          class="a11y-speak-region"
          id="a11y-speak-polite"
          style="position: absolute;margin: -1px;padding: 0;height: 1px;width: 1px;overflow: hidden;clip: rect(1px, 1px, 1px, 1px);-webkit-clip-path: inset(50%);clip-path: inset(50%);border: 0;word-wrap: normal !important;"
        />
        <div />
      </body>
```